tests/assembleCoaddTestUtils.py

```diff
@@ -0,0 +1,625 @@
+                 tract=0, patch="2,3", visit=100):
+        self.coaddName = coaddName
+        self.tract = tract
+        self.patch = patch
+        self.visit = visit
+        visitInfo = exposure.getInfo().getVisitInfo()
+        self.datasetTypes = (f"{coaddName}Coadd_directWarp", f"{coaddName}Coadd_psfMatchedWarp",
+                             f"{coaddName}Coadd_directWarp_sub", f"{coaddName}Coadd_psfMatchedWarp_sub")
+        self.metadataTypes = (f"{coaddName}Coadd_directWarp_visitInfo",
+                              f"{coaddName}Coadd_psfMatchedWarp_visitInfo")
+        self.dataLookup = {f"{coaddName}Coadd_directWarp": exposure,
+                           f"{coaddName}Coadd_psfMatchedWarp": exposurePsfMatched,
+                           f"{coaddName}Coadd_directWarp_sub": exposure,
+                           f"{coaddName}Coadd_psfMatchedWarp_sub": exposurePsfMatched,
+                           f"{coaddName}Coadd_directWarp_visitInfo": visitInfo,
+                           f"{coaddName}Coadd_psfMatchedWarp_visitInfo": visitInfo}
+
+    def get(self, datasetType, bbox=None, immediate=True):
+        """Retrieve the specified dataset using the API of the Gen 2 Butler.
+
+        Parameters
+        ----------
+        datasetType : `str`
+            Name of the type of exposure to retrieve.
+        bbox : `lsst.geom.box.Box2I`, optional
+            If supplied, retrieve only a subregion of the exposure.
+        immediate : `bool`, optional
+            Ignored. Supplied for compatibility with the Gen 2 Butler.
+
+        Returns
+        -------
+        `lsst.afw.image.Exposure` or `lsst.afw.image.VisitInfo`
+            Either the exposure or its metadata, depending on the datasetType.
+
+        Raises
+        ------
+        ValueError
+            If an unknown datasetType is supplied.
+        """
+        if datasetType in self.datasetTypes:
+            exp = self.dataLookup[datasetType].clone()
+            if "_sub" in datasetType:
+                return exp[bbox]
+            else:
+                return exp
+        elif datasetType in self.metadataTypes:
+            return self.dataLookup[datasetType]
+        else:
+            raise ValueError(f"Unknown datasetType {datasetType}. Must be one of {self.datasetTypes}.")
+
+    @property
+    def dataId(self):
+        """Generate a valid data identifier.
+
+        Returns
+        -------
+        dataId : `lsst.daf.persistence.dataId`
+            Data identifier dict for patch.
+        """
+        return DataId(tract=self.tract, patch=self.patch, visit=self.visit)
+
+    def datasetExists(self, tempExpName):
+        """Mimic the Gen2 Butler API for determining whether a dataset exists.
+        """
+        if tempExpName in self.datasetTypes:
+            return True
+        else:
+            return False
+
+
+class MockExposureReference(DeferredDatasetHandle):
+    """Very simple object that looks like a Gen 3 data reference to an exposure.
```

Comment:
It needs to inherit from `DeferredDatasetHandle` because the way `AssembleCoaddTask` differentiates between a Gen 2 or Gen 3 data reference is by testing `if isinstance(warpExpRef, DeferredDatasetHandle):`

Reply:
The name is because it is a data reference to an `lsst.afw.image.Exposure` object. Each data reference is also to a single warped exposure, and not actually anything coadded yet.

Reply:
That isinstance is unfortunate, but I guess that's on the Coadd code.